### PR TITLE
Support docker registry v2 push/pull

### PR DESCRIPTION
**Pull request by qiangli**
Implement docker registry v2 push/pull based on https://github.com/google/go-containerregistry
Refactored to use IPFS v1 cid

Fix issue #13
Fix issue #7

### CODE REVIEW

**Comment by miguelmota** [comment]
Thank you for the PR @qiangli! Going to review and merge

File: cmd/ipdr/main.go [hunk 1]
```diff
@@ -137,23 +141,38 @@ More info: https://github.com/miguelmota/ipdr`,
...
+		defaultCIDStore = filepath.Join(defaultCIDStore, ".ipdr/cids")
+	}
+
 	serverCmd.Flags().BoolVarP(&silent, "silent", "s", false, "Silent flag suppresses logs")
 	serverCmd.Flags().UintVarP(&port, "port", "p", 5000, "The port for the Docker registry to listen on")
 	serverCmd.Flags().StringVarP(&tlsCertPath, "tlsCertPath", "", "", "The path to the .crt file for TLS")
 	serverCmd.Flags().StringVarP(&tlsKeyPath, "tlsKeyPath", "", "", "The path to the .key file for TLS")
+	serverCmd.Flags().StringVarP(&ipfsHost, "ipfs-host", "", "127.0.0.1:5001", "A remote IPFS API host to pull the image from. Eg. 127.0.0.1:5001")
 	serverCmd.Flags().StringVarP(&ipfsGateway, "ipfs-gateway", "g", "127.0.0.1:8080", "The readonly IPFS Gateway URL to pull the image from. Eg. https://ipfs.io")
+	serverCmd.Flags().StringArrayVar(&cidResolvers, "cid-resolver", []string{"file:" + defaultCIDStore, "oci.dhnt.io"}, "Map repo:reference to CID. Accepts dnslink, IPFS path, and local file path.")
```
Comment (miguelmota) [inline cmd/ipdr/main.go:174]: What is `oci.dhnt.io`?

Comment (miguelmota) [reply cmd/ipdr/main.go:174]: What do you think about only having `[]string{"/ipfs"}` as the default here?

**Comment by miguelmota** [comment]
Tested and working. Amazing work @qiangli. The dig command is a great addition.

File: cmd/ipdr/main.go (hunk 1 above)
Comment (qiangli) [reply cmd/ipdr/main.go:174]: A domain  dedicated to  resolving dnslink (https://dnslink.io/ ) for publicly sharing docker image name[:tag] -> cid mapping on IPFS.
I own the tld.

Comment (miguelmota) [reply cmd/ipdr/main.go:174]: Ah I see.
Is the dnslink value a hash of the local cids directory? or how do you manage or update the cid resolver domain?
For example, this is how I tested it out:

Pull and tag hello world image:

```bash
docker pull hello-world
docker tag hello-world docker.local:5000/hello-world
docker push --quiet docker.local:5000/hello-world
```

Tree of cids directory contents:

```bash
$ tree ~/.ipdr/
/home/mota/.ipdr/
└── cids
    └── hello-world
        └── latest
```

Add cids directory to IPFS:

```bash
$ ipfs add -r ~/.ipdr/cids/ --quieter
QmVtjwa3kdFJHce2wnFuygaCHdSPXraBd4FRSEZKjqZWQp
```

Set _dnslink TXT record and verify:

```bash
$ dig _dnslink.ipdr.io -t TXT +short
"dnslink=/ipfs/QmVtjwa3kdFJHce2wnFuygaCHdSPXraBd4FRSEZKjqZWQp"
```

Run server with resolver domain:

```bash
go run cmd/ipdr/ipdr.go server --cid-resolver=ipdr.io
```

Run ipdr dig to get cid using repo tag name:

```bash
$  ipdr dig hello-world:latest
bafybeiakvswzlopeu573372p5xry47tkc2hhcg5q5rulmbfrnkecrbnt3y
```

We could use `ipdr.io` domain as a canonical resolver

Comment (qiangli) [reply cmd/ipdr/main.go:174]: Yes, it is the v1 hash of the top directory. 
`ipfs add -r --cid-version 1 cids`

Comment (qiangli) [reply cmd/ipdr/main.go:174]: May I also suggest resolve ` local.ipdr.io`  (or similar) to localhost host address `127.0.0.1` as a convenience?
This way we could skip the step of editing the /etc/hosts.

Comment (miguelmota) [reply cmd/ipdr/main.go:174]: @qiangli good idea. It's supported now:

```bash
$ dig local.ipdr.io +short
127.0.0.1
```

Comment (miguelmota) [reply cmd/ipdr/main.go:174]: Moved the repository to an organization: https://github.com/ipdr/ipdr

You should receive an invite to be a collaborator on repo : )

@qiangli

Comment (qiangli) [reply cmd/ipdr/main.go:174]: thanks!

**Pull request merged**